model/semester.go

```diff
@@ -0,0 +1,33 @@
+package model
+
+type Semester struct {
+	TeachingTerm string
+	Year         int
+}
+
+// InRangeOfSemesters checks if s is between firstSemester (inclusive) and lastSemester (inclusive) or is element of semesters slice
+func (s *Semester) InRangeOfSemesters(firstSemester Semester, lastSemester Semester, semesters []Semester) bool {
+	if s == nil {
+		return false
+	}
+	if semesters == nil {
```

Comment:
given that this quite subtely splits the controlflow (`firstSemester`/`lastSemester` is only checked if no `semesters` exist, which would not be obvious from the docsstring), using two functions for the different functionality where previously `InRangeOfSemesters` is used would make sense.

Reply:
The straightforward solution in other programming languages would be to use method overloading. Since this is not possible in Go, I brought this up in our weekly meeting a while ago and iirc agreed on this solution. 
Also consider that a change in this method would propagate upwards through the usage chain and require changes not only in this method + adding more complexity to more methods than just this one. 

From this background, is adding clarifying comments enough? Is there any best practice in Go I'm not aware of for legitimate use cases of method overloading?

Reply:
I am going to give you a bit of pushback on mushing 2 methods together as good style just because java somewhat does that.
Using overloading like hides the controllflow a lot.

Especially that the `firstSemester`/`lastSemester` parameter is unused for `InRangeOfSemesters` if `semesters` is set is likely not obvious..
I don't think the usage of this function understand this nuance.

Having them be one function, does loose the quite subtle difference

This is essentially these two functions (python to better reason about this):

```py
@dataclasses.dataclass # to get equality and constructors
class Semester:
  term: Literal["WS","SS"]
  term: int

  def in_range_of_semesters(self, semesters:  List[Semester]):
    return any(s == self for s in semesters)

  def in_range_of(first_inclusive: Semester, end_inclusive: Semester):
    direct_match = self == first_inclusive or self == end_inclusive
    semester_is_between_limits = self.is_after(first_inclusive) and not self.is_after(end_inclusive)
    return direct_matchs or semester_is_between_limits

  def is_after(self, other: Semester):
    same_year_but_term_after = other.year == self.year and self.term == "WS" and  self.term == "SS"
    return self.year > other.year or same_year_but_term_after
```

> [!TIP]
> in the end, my word does not count in this project.
> I have not put in the work
> => feel free to ignore 😉

Reply:
I generally agree with what you see as problematic.

I did split the function into two (in the calling method as well). There are still parts, where this still is kind of mushed together, but only in non-exported functions in api/search.go, where I added comments. => Still far from optimal, but I guess, it's better when that is only the case in those parts of the code, where one probably has to look and understand the code anyway before doing something with it.